src/mpi/coll/reduce/reduce.c

```diff
@@ -321,6 +334,94 @@ int MPIR_Reduce_impl(const void *sendbuf, void *recvbuf, int count,
     goto fn_exit;
 }
 
+#undef FUNCNAME
+#define FUNCNAME MPIR_Reduce_decision_tree
+#undef FCNAME
+#define FCNAME MPL_QUOTE(FUNCNAME)
+int MPIR_Reduce_decision_tree(const void *sendbuf, void *recvbuf, int count, MPI_Datatype datatype,
+                              MPI_Op op, int root, MPIR_Comm * comm_ptr, MPIR_Errflag_t * errflag)
+{
+    int mpi_errno = MPI_SUCCESS;
+    size_t size;
+    cJSON *device_info;
+
+    /* TODO - Calculate size of buffer */
+
+    mpi_errno = MPIU_Decision_tree_traverse("reduce", comm_ptr, size, datatype, coll_info);
+
+    if (unlikely(mpi_errno != MPI_SUCCESS))
+        MPIR_ERR_POP(mpi_errno);
+    else {
```

Comment:
Should this be in `else`? Doesn't `MPIR_ERR_POP` jump to `fn_fail` in case of a failure?

Reply:
The idea was to hide all decistion making on _select level. Here we blend decision tree logic (it could be n dimentional array) with collective call logic.

The _select function returns container which is a coll_id + its parameters